config/config.go

```diff
@@ -161,6 +166,174 @@ type AuthConfig struct {
+
+	// All contains a list of AuthenticationRequirements.
+	// A successful authentication requires all of them being satisfied.
+	All []AuthenticationRequirement `yaml:"all,omitempty" json:"all,omitempty"`
+}
+
+// JWTAuthentication defines the JWT authentication
+type JWTAuthentication struct {
+	// Name of this JWT Provider, unique within the Proxy.
+	Name string `yaml:"name" json:"name"`
+
+	// JWT issuer ("iss" claim)
+	Issuer string `yaml:"issuer" json:"issuer"`
+
+	// A remote JWKS source
+	RemoteJWKS RemoteJWKS `yaml:"remote_jwks" json:"remote_jwks"`
+
+	// Audiences contains a list of audiences
+	Audiences []string `yaml:"audiences,omitempty" json:"audiences,omitempty"`
+
+	// Header name that will contain decoded JWT payload in requests forwarded to
+	// target.
+	ForwardPayloadHeader string `yaml:"forward_payload_header,omitempty" json:"forward_payload_header,omitempty"`
+
+	// Locations where JWT may be found. First match wins.
+	In []HTTPParameter `yaml:"in" json:"in"`
+}
+
+// RemoteJWKS contains information for remote JWKS
+type RemoteJWKS struct {
+	// URL of the JWKS
+	URL string `yaml:"url" json:"url"`
+
+	// CacheDuration of the JWKS
+	CacheDuration time.Duration `yaml:"cache_duration,omitempty" json:"cache_duration,omitempty"`
+}
+
+// ConsumerAuthorization is the configuration of API consumer authorization
+type ConsumerAuthorization struct {
+	FailOpen bool `yaml:"fail_open,omitempty" json:"fail_open,omitempty"`
+
+	// Locations of API consumer credential (API Key). First match wins.
+	In []HTTPParameter `yaml:"in" json:"in"`
+}
+
+// Target contains the name of the target server and request transformation
+type Target struct {
+	// Name of the target
+	Name string `yaml:"name" json:"name"`
+
+	// RequestTransformation defines how the request should be modified, e.g. a path rewrite
+	RequestTransformation RequestTransformation `yaml:"request_transformation,omitempty" json:"request_transformation,omitempty"`
+}
+
+type RequestTransformation struct {
+	// StringTransformation to rewrite the path request header
+	PathRewrite StringTransformation `yaml:"path_rewrite,omitempty" json:"path_rewrite,omitempty"`
+}
+
+// HTTPMatch is an HTTP request matching rule
+type HTTPMatch struct {
+	// URL path template using to match incoming requests and optionally identify
+	// path variables.
+	PathTemplate string `yaml:"path_template" json:"path_template"`
+
+	// HTTP method (e.g. GET, POST, PUT, etc.)
+	Method string `yaml:"method,omitempty" json:"method,omitempty"`
+}
+
+// HTTPParameter defines an HTTP paramter
+// Precisely one of Query, Header or JWTClaim should be set.
+type HTTPParameter struct {
+	// Name of a query paramter
+	Query string `yaml:"query,omitempty" json:"query,omitempty"`
+
+	// Name of a header
+	Header string `yaml:"header,omitempty" json:"header,omitempty"`
+
+	// A JWTClaim
+	JWTClaim JWTClaim `yaml:"jwt_claim,omitempty" json:"jwt_claim,omitempty"`
```

Comment:
These three fields are mutually exclusive, are we planning on writing validation for this? JFYI there is an alternative which is the use pseudo-union type and a custom YAML/JSON marshaller impl, which makes the usage of the config much more straightforward via type switch e.g.

```
type ParamMatch interface {
  paramMatch()
}

type HTTPParameter struct {
  Match ParamMatch
  Transformation StringTransformation
}

type Query string
func (Query)  paramMatch() {}

type Header string
func (Header)  paramMatch() {}

type JWTClaim struct {...}
func (JWTClaim)  paramMatch() {}

func (p *HTTPParameter) UnmarshalJSON(b []byte) error {
  // implement custom logic for the Match field
}

func exampleUsage() {
  p := &HTTPParameter{}
  _ := json.Unmarshal(p, data)
  switch match := p.Match.(type) {
    case Query:
      ...
    case Header:
      ...
    case JWTClaim:
      ...
}
```

Reply:
Ah yeah that will be a good idea. The same thing can be done for the `JWKSSource` too.